### PR TITLE
Proof-of-concept draft for teleport-update binary implementation

### PR DESCRIPTION
**Pull request by sclevine**
This PR contains an initial draft of the `teleport-update` binary described in https://github.com/gravitational/teleport/pull/40190.

Not all functionality is implemented, only:
- Atomic downloading and extraction of the Teleport tgz to `/var/lib/teleport/versions/X.Y.Z`, with disk space checks, checksum verification, and protection against power loss scenarios.
- Atomic configuration management of `updates.yaml`

This PR will not be merged as-is, and is only open for early review of the approach.

I plan to break this download into separate, tested PRs if this initial version looks good.

```shell
root@71aeb39eabec:/# ./teleport/teleport-update enable --proxy levine.teleport.sh --debug
2024-09-06T20:09:50Z DEBU  Attempting request to Proxy web api method:GET host:levine.teleport.sh path:/webapi/find webclient/webclient.go:134
2024-09-06T20:09:51Z INFO [UPDATER]   Downloading Teleport tarball path:/tmp/teleport-update-1578638652 size:153512064 teleport-update/versions.go:210
2024-09-06T20:10:01Z INFO  extracted tarball into /var/lib/teleport/versions/16.2.0: 582 files, 93 dirs (2.678702627s)
```

### CODE REVIEW

File: api/client/webclient/webclient.go [hunk 1]
```diff
@@ -171,6 +174,10 @@ func Find(cfg *Config) (*PingResponse, error) {
 
 	endpoint := fmt.Sprintf("https://%s/webapi/find", cfg.ProxyAddr)
 
+	if cfg.Group != "" {
+		endpoint = fmt.Sprintf("%s?group=%s", endpoint, cfg.Group)
+	}
+
```
[the remark this answers is not in the document]
Comment (sclevine) [reply api/client/webclient/webclient.go:180]: Sorry, totally agree, was just using this to test

File: tool/teleport-update/versions.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+	// Ensure there's enough space in /tmp for the download.
+	size := resp.ContentLength
+	if size < 0 {
+		plog.Warn("Content length missing from response, unable to verify Teleport download size")
+	} else if size > free {
+		return nil, nil, trace.Errorf("size of download (%d bytes) exceeds available disk space (%d bytes)", resp.ContentLength, free)
+	}
+
+	// Calculate checksum concurrently with download.
```
[the remark this answers is not in the document]
Comment (sclevine) [reply tool/teleport-update/versions.go:220]: No downside to it, and it avoids needed to re-read the file from disk

File: tool/teleport-update/versions.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+		return nil, trace.Wrap(err)
+	}
+	raw := buf.String()
+	if n != 64 {
+		plog.WarnContext(ctx, "unexpected checksum size", "size", n, "checksum", raw)
+	}
+	sum, err := hex.DecodeString(raw)
+	if err != nil {
+		plog.WarnContext(ctx, "corrupt checksum detected", "size", n, "checksum", raw)
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (sclevine) [reply tool/teleport-update/versions.go:178]: Agree, will add a typed error in the separate PR for this function.

File: tool/teleport-update/versions.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+	}
+	defer f.Close()
+	var buf bytes.Buffer
+	n, err := io.Copy(&buf, io.LimitReader(f, 65))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	raw := buf.String()
+	if n != 64 {
+		plog.WarnContext(ctx, "unexpected checksum size", "size", n, "checksum", raw)
```
[the remark this answers is not in the document]
Comment (sclevine) [reply tool/teleport-update/versions.go:173]: Will cover this with a typed error (same as below)

[the remark this answers is not in the document]
Comment (vapopov) [reply tool/teleport-update/versions.go:173]: nit: might be replaced with `io.CopyN` which actually does the same, and returns `EOF` if n is less than expected

**Comment by sclevine** [comment]
Completed by https://github.com/gravitational/teleport/pull/47565

**Pull request closed without merging**